File: web/src/app/model/dashboard-column.ts
import {Attribute} from "./attribute";

export class DashboardColumn {
    id: number;

    attribute: Attribute;

    order: Order = 'NONE';

    position: number = 9999;

    width: number = 200;

    widthDelta: number = 0;

    getFullWidth() {
        return this.width + this.widthDelta;
    }

    static fromAttribute(attribute: Attribute): DashboardColumn {
        let column = new DashboardColumn();
        column.attribute = attribute;
        return column;
    }

    static from(value: any): DashboardColumn {
        let column = new DashboardColumn();
        column.id = value.id;
        column.attribute = Attribute.from(value.attribute);
        column.order = value.order;
        column.position = value.position;
        column.width = value.width;
        return column;
    }

    toSave(): any {
        return {
            id: this.id,
            attribute: this.attribute.toSave(),
            order: this.order,
            position: this.position,
            width: this.width
        }
    }

}

type Order = 'ASC' | 'DESC' | 'NONE'

File: server/src/main/java/com/evgenltd/lt/entity/DashboardColumn.java
package com.evgenltd.lt.entity;

import javax.persistence.*;

@Entity
@Table(name = "dashboard_columns")
public class DashboardColumn {

    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    private Long id;

    @ManyToOne
    @JoinColumn(name = "dashboard_id")
    private Dashboard dashboard;

    @ManyToOne
    @JoinColumn(name = "attribute_id")
    private Attribute attribute;

    @Enumerated(EnumType.STRING)
    private Order order;

    private Integer position;

    private Integer width;

    public Long getId() {
        return id;
    }

    public void setId(final Long id) {
        this.id = id;
    }

    public Dashboard getDashboard() {
        return dashboard;
    }

    public void setDashboard(final Dashboard dashboard) {
        this.dashboard = dashboard;
    }

    public Attribute getAttribute() {
        return attribute;
    }

    public void setAttribute(final Attribute attribute) {
        this.attribute = attribute;
    }

    public Order getOrder() {
        return order;
    }

    public void setOrder(final Order order) {
        this.order = order;
    }

    public Integer getPosition() {
        return position;
    }

    public void setPosition(final Integer position) {
        this.position = position;
    }

    public Integer getWidth() {
        return width;
    }

    public void setWidth(final Integer width) {
        this.width = width;
    }

    public enum Order {
        NONE,
        ASC,
        DESC
    }

}
